### PR TITLE
rpf check

### PR DESCRIPTION
**Pull request by tomastigera**
## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: bpf-gpl/tc.c [hunk 1]
```diff
@@ -535,11 +537,28 @@ static CALI_BPF_INLINE int calico_tc(struct __sk_buff *skb)
...
+	/* We are possibly past (D)NAT, but that is ok, we need to let the IP
+	 * stack do the RPF check on the source, dest is not importatnt.
+	 */
+	if (CALI_F_TO_HOST && ct_result_rpf_failed(state.ct_result.rc)) {
+		fwd_fib_set(&fwd, false);
+	}
+
+
 	/* skip policy if we get conntrack hit */
 	if (state.ct_result.rc != CALI_CT_NEW) {
```
Comment (fasaxc) [inline bpf-gpl/tc.c:549]: Does this check need to mask the rc too?

File: fv/bpf_test.go [hunk 2]
```diff
@@ -617,6 +617,160 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
 					cc.CheckConnectivity()
 				})
 
+				It("should not be able to spoof workloads IP by another workload", func() {
+					if testOpts.protocol != "udp" {
+						return
+					}
+
+					felixes[0].EnsureBinary("pktgen")
```
Comment (fasaxc) [inline fv/bpf_test.go:625]: Could just build all the test binaries into the fat docker image now

File: fv/bpf_test.go [hunk 3]
```diff
@@ -617,6 +617,160 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
...
+
+					By("testing that packet sent by another workload is dropped", func() {
+						tcpdump := w[0][0].AttachTCPDump()
+						tcpdump.SetLogEnabled(true)
+						matcher := fmt.Sprintf("IP %s.30444 > %s.30444: UDP", w[1][0].IP, w[0][0].IP)
+						tcpdump.AddMatcher("UDP-30444", regexp.MustCompile(matcher))
+						tcpdump.Start(testOpts.protocol, "port", "30444", "or", "port", "30445")
+						defer tcpdump.Stop()
+
+						// send a packet from the correct workload to create a conntrac entry
```
Comment (fasaxc) [inline fv/bpf_test.go:662]: Typo: conntrac

File: fv/bpf_test.go [hunk 4]
```diff
@@ -617,6 +617,160 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
...
+					 */
+					By("setting strict RPF check", func() {
+						felixes[1].Exec("sysctl", "-w", "net.ipv4.conf.all.rp_filter=1")
+						felixes[1].Exec("sysctl", "-w", "net.ipv4.conf.default.rp_filter=1")
+					})
+
+					By("testing that packet sent by another workload is dropped", func() {
+						tcpdump := w[0][0].AttachTCPDump()
+						tcpdump.SetLogEnabled(true)
+						matcher := fmt.Sprintf("IP %s.30444 > %s.30444: UDP", w[1][0].IP, w[0][0].IP)
```
Comment (fasaxc) [inline fv/bpf_test.go:657]: This is test code, so not super important but the IP and the format string both have `.` in them, which is a wildcard.  I think there's `regexp.QuoteMeta()` to do escaping where you want a literal match

File: fv/bpf_test.go [hunk 5]
```diff
@@ -617,6 +617,160 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
...
+						tcpdump.AddMatcher("UDP-30444", regexp.MustCompile(matcher))
+						tcpdump.Start(testOpts.protocol, "port", "30444", "or", "port", "30445")
+						defer tcpdump.Stop()
+
+						// send a packet from the correct workload to create a conntrac entry
+						_, err := w[1][0].RunCmd("/pktgen", w[1][0].IP, w[0][0].IP, "udp",
+							"--port-src", "30444", "--port-dst", "30444")
+						Expect(err).NotTo(HaveOccurred())
+
+						// We must eventually se the packet at the target
```
Comment (fasaxc) [inline fv/bpf_test.go:667]: Typoe: `se`

File: fv/bpf_test.go [hunk 6]
```diff
@@ -617,6 +617,160 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
 					cc.CheckConnectivity()
 				})
 
+				It("should not be able to spoof workloads IP by another workload", func() {
```
Comment (fasaxc) [inline fv/bpf_test.go:620]: If useful, you can add a config param to felix prefixed with `Debug...` to disable the RPF check in  order to cross-check that the only thing dropping the packet is the new check.  We have a few of those for testing normally-unhittable cases, such as `DebugSimulateDataplaneHangAfter`

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looking great so far; especially the new packet generation stuff for the tests.

Couple of things that I think you still need to do:

- [ ] Make sure felix is setting up the RPF check and remove from the FV.
- [ ] Rename the conntrack map so that we won't clash on upgrade.  I think we can tolerate a breaking change like that before we go GA.
- [ ] Remove the accidentally checked in binary files

File: bpf-gpl/tc.c (hunk 1 above)
Comment (tomastigera) [reply bpf-gpl/tc.c:549]: technically no, but I will do it, good point

File: fv/bpf_test.go (hunk 2 above)
Comment (tomastigera) [reply fv/bpf_test.go:625]: done

File: fv/bpf_test.go (hunk 6 above)
Comment (tomastigera) [reply fv/bpf_test.go:620]: good point, probably does not matter here, I escaped those manually

**Pull request closed without merging**